File: ff_flag_tracker.py
class FlagState:
    def __init__(self):
        self.at_base = True
        self.carried = False
        self.dropped = False
        self.carrier = 0

class FFFlagTracker:
    def __init__(self):
        self.flags = {2: FlagState(), 3: FlagState(), 4: FlagState(), 5: FlagState()}

    def flag_picked_up(self, team, userid):
        f = self.flags[team]
        f.at_base = False
        f.carried = True
        f.dropped = False
        f.carrier = userid

    def flag_returned(self, team):
        f = self.flags[team]
        f.at_base = True
        f.carried = False
        f.dropped = False
        f.carrier = 0

    def flag_captured(self, team):
        self.flag_returned(team)

    def get_flag(self, team):
        return self.flags[team]

File: utils/RCBot2_meta/ff_flag_tracker.cpp
#include "ff_flag_tracker.h"
#include "engine_wrappers.h"

CFFFlagTracker g_FFFlagTracker;

CFFFlagTracker::CFFFlagTracker()
{
    Reset();
}

void CFFFlagTracker::Reset()
{
    for(int i=0;i<6;i++)
        m_flags[i] = FlagState();
}

void CFFFlagTracker::FlagPickedUp(int team, int userid)
{
    if(team<0||team>=6) return;
    m_flags[team].atBase = false;
    m_flags[team].carried = true;
    m_flags[team].dropped = false;
    m_flags[team].carrier = userid;
}

void CFFFlagTracker::FlagReturned(int team)
{
    if(team<0||team>=6) return;
    m_flags[team] = FlagState();
}

void CFFFlagTracker::FlagCaptured(int team)
{
    FlagReturned(team);
}

const FlagState &CFFFlagTracker::GetFlag(int team) const
{
    static FlagState dummy;
    if(team<0||team>=6) return dummy;
    return m_flags[team];
}
